aten/src/ATen/native/mobile/cpu/Engine.h

```diff
@@ -0,0 +1,137 @@
+#pragma once
+
+#include <ATen/ATen.h>
+
+// Prefer not to use these functions directly.  Instead prefer to use the
+// corresponding mobile operators that this module registers with, and exposes
+// through, c10 whenever possible in order to decouple op creation from op
+// execution.  The underlying computation in many NN operators is performed in
+// phases, all of which do not need to run at the same frequency.  Decoupling op
+// creation from execution allows for one-time computations to be cached and
+// factored out resulting in time savings per calls to forward() - something
+// this API cannot do.  Furthermore, this API does not allow for fusion of
+// non-linear operators that, again, is something that the exposed c10 mobile
+// operators can handle.
+
+namespace at {
+namespace native {
+namespace mobile {
+namespace cpu {
+
+bool available();
+bool initialize();
+bool deinitialize();
```

Comment:
I guess I will find out when I read the rest of this patch, but I very much want to know what these actually do

Reply:
Initialize and deinitialize XNNPACK.  We are doing a similar thing for NNPACK as well.

Reply:
The situation with NNPACK is a bit different. Initialization of NNPACK is private global state, which is not made visible to clients of the functions of NNPACK. One way to see this is `init_nnpack` is a static function in a C++ file. However, in your patch, you've put the initialization/deinitalization in a header...

Reply:
I can remove these functions from this file, but still need to expose it in an internal header or manually declare it in all files where it is used.  The difference between XNNPACK and NNPACK integration is that we are only using NNPACK for convolutions, so one local function definition is enough there, while we need to call this function from each operator in this directory considering that we are using XNNPACK for more than convolutions.  The performance difference of the non-convolution operators between XNNPACK and native::add etc is still considerable so this is the path we want to hit on mobile.

Reply:
OK, fair enough!

Reply:
(You don't have to move it.)